lib/core/test/unit/Cardano/Wallet/DB/Properties.hs

```diff
@@ -862,67 +896,173 @@ prop_rollbackTxHistory db@DBLayer{..} (InitialCheckpoint cp0) (GenTxHistory txs0
+-- a sequence of blocks, the following steps:
+--
+--  For all batch B in sequence:
+--
+--  - Keep all checkpoints in B yielded by sparseCheckpoint with and
+--    edge size of 0.
+--
+--  - Keep the last checkpoint of the batch regardless
+--
+--  - Prune all checkpoints not yielded by sparseCheckpoint with a non-null edge
+--    size
+--
+-- are equivalent to calling `sparseCheckpoints` on the flattened batch
+-- sequence.
+--
+-- Note that the batch creation mimics the way blocks are served by the network
+-- layer to wallets: first by batches of arbitrary size, and then one by one.
+--
+-- The property shows that regardless of how batches are served, after
+-- 'edgeSize' one-by-one step, the end result is the same as if the entire
+-- stream of blocks had been processed in one go.
+prop_checkpointsEventuallyEqual
+    :: GenSparseCheckpointsArgs
+    -> Property
+prop_checkpointsEventuallyEqual args@(GenSparseCheckpointsArgs cfg (k,h)) = prop
+    & counterexample ("h=" <> show h)
+    & counterexample ("k=" <> show k)
+  where
+    prop :: Property
+    prop = forAll (genBatches args) $ \(Batches batches) ->
+        let
+            tip =
+                Quantity $ last $ mconcat batches
+            emptyDB =
+                SparseCheckpointsDB []
+            SparseCheckpointsDB db =
+                L.foldr (\batch -> prune . step batch) emptyDB batches
+        in
+            db === sparseCheckpoints cfg (Quantity k) tip
+
+    step :: [Word32] -> SparseCheckpointsDB -> SparseCheckpointsDB
+    step cps (SparseCheckpointsDB db) =
+        let
+            toKeep =
+                sparseCheckpoints (cfg { edgeSize = 0 }) (Quantity k) (Quantity h)
+            cps' =
+                last cps : (toKeep `L.intersect` cps)
+        in
+            SparseCheckpointsDB $ L.sort $ cps' ++ db
+
+    prune :: SparseCheckpointsDB -> SparseCheckpointsDB
+    prune (SparseCheckpointsDB db) =
+        let
+            tip =
+                Quantity $ last db
+            db' =
+                sparseCheckpoints cfg (Quantity k) tip `L.intersect` db
+        in
+            SparseCheckpointsDB db'
+
+newtype Batches = Batches [[Word32]] deriving Show
+
+newtype SparseCheckpointsDB = SparseCheckpointsDB [Word32] deriving Show
+
 int :: Integral a => a -> Int
 int = fromIntegral
 
 pp :: ProtocolParameters
 pp = dummyProtocolParameters
 
-newtype GenSparseCheckpointsArgs
-    = GenSparseCheckpointsArgs (Word32, Word32)
-    deriving newtype Show
+data GenSparseCheckpointsArgs
+    = GenSparseCheckpointsArgs SparseCheckpointsConfig (Word32, Word32)
+    deriving Show
 
 instance Arbitrary GenSparseCheckpointsArgs where
     arbitrary = do
         k <- (\x -> 10 + (x `mod` 1000)) <$> arbitrary
         h <- (`mod` 100000) <$> arbitrary
-        pure $ GenSparseCheckpointsArgs ( k, h )
+        cfg <- SparseCheckpointsConfig
+            <$> choose (1, k-1)
+            <*> choose (0, 10)
```

Comment:
So we can only expect the properties above to hold in the real world, if the real world conforms to these conditions…

Would it be possible to move these conditions to some other place? Like
1. Explicit preconditions in the properties
2. Precondition when constructing a `SparseCheckpointsConfig`

On the same note, you mentioned "So this works so long as the network batch size remains smaller than k". I suppose that would be good to mention in the documentation for `SparseCheckpointsConfig` or `sparseCheckpoints`?

Reply:
> 1. Explicit preconditions in the properties

I disagree, that's not how pre-conditions should be used in property tests. 

> 2. Precondition when constructing a SparseCheckpointsConfig

This sounds more reasonable. Either as invariants or with a smart constructor. 

> I suppose that would be good to mention in the documentation for SparseCheckpointsConfig or sparseCheckpoints?

Agree, will do.

Reply:
> that's not how pre-conditions should be used in property tests

I'm not convinced entirely convinced. I've tried to raise a question: 
https://input-output-rnd.slack.com/archives/C48HA39RS/p1600345143060900